Per channel fake quant

raghuramank100 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26627 Default observer and fake-quant for backends
* #26625 Emulate weight and activation only quant with fake quant, numerics test
* #26624 Quantization aware training: Freeze batch norm support
* **#26623 Per channel fake quant**

Per-channel fake quant cpu and cuda operators,
per-channel support in fake quant module,
tests for per-channel fake-quant and serializability of fake quant modules


Differential Revision: [D17439406](https://our.internmc.facebook.com/intern/diff/D17439406/)

raghuramank10000 committed:
Per channel fake quant

Per-channel fake quant cpu and cuda operators,
per-channel support in fake quant module,
tests for per-channel fake-quant and serializability of fake quant modules

ghstack-source-id: 90483681

Differential Revision: [D17439406](https://our.internmc.facebook.com/intern/diff/D17439406/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Per channel fake quant"


Per-channel fake quant cpu and cuda operators,
per-channel support in fake quant module,
tests for per-channel fake-quant and serializability of fake quant modules


Differential Revision: [D17439406](https://our.internmc.facebook.com/intern/diff/D17439406/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Per channel fake quant"


Per-channel fake quant cpu and cuda operators,
per-channel support in fake quant module,
tests for per-channel fake-quant and serializability of fake quant modules


Differential Revision: [D17439406](https://our.internmc.facebook.com/intern/diff/D17439406/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Per channel fake quant"


Per-channel fake quant cpu and cuda operators,
per-channel support in fake quant module,
tests for per-channel fake-quant and serializability of fake quant modules


Differential Revision: [D17439406](https://our.internmc.facebook.com/intern/diff/D17439406/)

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/fake_quantize_per_channel_affine.cpp:
```diff
@@ -0,0 +1,141 @@
+      "`quant_min` should be less than or \
+        equal to `quant_max`.");
+
+//      return self;
```
remove?

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/fake_quantize_per_channel_affine.cpp:
```diff
@@ -0,0 +1,141 @@
+#      "`zero_point` must be between `quant_min` and `quant_max`.");
+*/
+  auto Y = at::empty_like(self);
+  for (int i = 0; i < self.size(0); i++)
```
assert axis == 0?

dzhulgakov replied:
or better yet - just use `axis` - from the code below it seems that it's trivial to do since you're using slice

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/fake_quantize_per_channel_affine.cpp:
```diff
@@ -0,0 +1,141 @@
+    float inv_scale = 1.0f / sc;
+    int64_t z_point = zero_point[i].item().toLong();
+    auto iter = TensorIterator::unary_op(output_slice, input_slice);
+    cpu_kernel(iter, [&](float self) -> float {
```
should we extract this into a separate function so that we can be sure that fq_per_channel and fq_per_tensor are using the same thing?

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/fake_quantize_per_channel_affine.cpp:
```diff
@@ -0,0 +1,141 @@
+        equal to `quant_max`.");
+
+//      return self;
+/*  TORCH_CHECK(
```
change the check to work with Tensor?

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/fake_quantize_per_channel_affine.cpp:
```diff
@@ -0,0 +1,141 @@
+      quant_min <= quant_max,
+      "`quant_min` should be less than or \
+        equal to `quant_max`.");
+  /*TORCH_CHECK(
```
change the check to work with Tensor?

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/fake_quantize_per_channel_affine.cpp:
```diff
@@ -0,0 +1,141 @@
+
+  Tensor dX = at::zeros_like(X);
+
+  for (int i = 0; i < X.size(0); i++)
```
assert for axis here as well

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/fake_quantize_per_channel_affine.cpp:
```diff
@@ -0,0 +1,141 @@
+    int64_t z_point = zero_point[i].item().toLong();
+    auto iter = TensorIterator::binary_op(ZdX, ZX, ZdY);
+    cpu_kernel(iter, [&](float x, float dy) -> float {
+    int64_t xq =
```
same here, should we extract this as a separate function that's used by both fq ops?

jerryzh168 commented on aten/src/ATen/native/quantized/cuda/fake_quantize_per_channel_affine.cu:
```diff
@@ -0,0 +1,136 @@
+      quant_min <= quant_max,
+      "`quant_min` should be less than or \
+        equal to `quant_max`.");
+/*  TORCH_CHECK(
```
check here as well

jerryzh168 reviewed (changes requested):
Could you fix the checks for arguments and formatting in tests?

raghuramank10000 committed:
Update on "Per channel fake quant"


Per-channel fake quant cpu and cuda operators,
per-channel support in fake quant module,
tests for per-channel fake-quant and serializability of fake quant modules


Differential Revision: [D17439406](https://our.internmc.facebook.com/intern/diff/D17439406/)

[ghstack-poisoned]

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/fake_quantize_per_channel_affine.cpp:
```diff
@@ -0,0 +1,141 @@
+*/
+Tensor fake_quantize_per_channel_affine_cpu(
+    const Tensor& self,
+    const Tensor& scale,
```
you need to add checks on scale/zp being 1D tensors

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/fake_quantize_per_channel_affine.cpp:
```diff
@@ -0,0 +1,141 @@
+#      zero_point >= quant_min && zero_point <= quant_max,
+#      "`zero_point` must be between `quant_min` and `quant_max`.");
+*/
+  auto Y = at::empty_like(self);
```
this defaults to contiguous memory layout, you need to explicitly pass memory_format = preserve if you want to keep NHWC (though not sure it's strictly needed)

raghuramank100 replied:
Its not really needed here as we operate in floating point, but seems like a good idea to not change layouts, will do that.

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/fake_quantize_per_channel_affine.cpp:
```diff
@@ -0,0 +1,141 @@
+    auto input_slice = self.slice(0,i,i+1);
+    auto output_slice = Y.slice(0,i,i+1);
+
+    float sc = scale[i].item().toFloat();
```
you could use TensorAccessor (it's faster as indexing creating a tensor). But it matters less because you have slicing above that creates tensors anyway

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/fake_quantize_per_channel_affine.cpp:
```diff
@@ -0,0 +1,141 @@
+  zero_point: zero_point of per tensor affine quantization
+  quant_min: minimum quantized value
+  quant_max: maximum quantized value
+  quant_delay: Count of global steps for which to delay the quantization.
```
looks like an outdated comment

dzhulgakov commented on test/common_quantized.py:
```diff
@@ -63,6 +63,35 @@ def _calculate_dynamic_qparams(X, dtype):
+    if isinstance(X, torch.Tensor):
+        X = X.numpy()
+    if dtype == torch.qint8:
+        qmin, qmax = -128, 127
```
you can use `torch.iinfo(dtype`

raghuramank10000 committed:
Update on "Per channel fake quant"


Per-channel fake quant cpu and cuda operators,
per-channel support in fake quant module,
tests for per-channel fake-quant and serializability of fake quant modules


Differential Revision: [D17439406](https://our.internmc.facebook.com/intern/diff/D17439406/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Per channel fake quant"


Per-channel fake quant cpu and cuda operators,
per-channel support in fake quant module,
tests for per-channel fake-quant and serializability of fake quant modules


Differential Revision: [D17439406](https://our.internmc.facebook.com/intern/diff/D17439406/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Per channel fake quant"


Per-channel fake quant cpu and cuda operators,
per-channel support in fake quant module,
tests for per-channel fake-quant and serializability of fake quant modules


Differential Revision: [D17439406](https://our.internmc.facebook.com/intern/diff/D17439406/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Per channel fake quant"


Per-channel fake quant cpu and cuda operators,
per-channel support in fake quant module,
tests for per-channel fake-quant and serializability of fake quant modules


Differential Revision: [D17439406](https://our.internmc.facebook.com/intern/diff/D17439406/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Per channel fake quant"


Per-channel fake quant cpu and cuda operators,
per-channel support in fake quant module,
tests for per-channel fake-quant and serializability of fake quant modules


Differential Revision: [D17439406](https://our.internmc.facebook.com/intern/diff/D17439406/)

[ghstack-poisoned]

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/fake_quantize_per_channel_affine.cpp:
```diff
@@ -0,0 +1,148 @@
+  auto Y = at::empty_like(self, self.options(), MemoryFormat::Preserve);
+  for (int i = 0; i < self.size(axis); i++)
+  {
+    auto input_slice = self.slice(axis,i,i+1);
```
we could rewrite it without slices (e.g. by passing the other tensor to tensoriterator and treating it as a binary operation with broadcasting) but if it's fast enough for you purposes - let's keep it for now

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/fake_quantize_per_channel_affine.cpp:
```diff
@@ -0,0 +1,148 @@
+    return X;
+  }
+
+  Tensor dX = at::zeros_like(X);
```
you need to preserve format here too. And why zeros? looks like you populate it fully anyway. It could be `empty_like`

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/fake_quantize_per_tensor_affine.cpp:
```diff
@@ -100,13 +86,13 @@ Tensor fake_quantize_per_tensor_affine_backward_cpu(
   }
 
   Tensor dX = at::zeros_like(X);
```
actually the same question about zeros

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/fake_quantize_per_channel_affine.cpp:
```diff
@@ -0,0 +1,148 @@
+    int64_t quant_max) {
+
+  TORCH_CHECK(self.scalar_type() == ScalarType::Float);
+
```
maybe check `scale.dim() == 1` and `zero_point.dim() == 1` as well

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/fake_quantize_per_channel_affine.cpp:
```diff
@@ -0,0 +1,148 @@
+
+  TORCH_CHECK(self.scalar_type() == ScalarType::Float);
+
+  TORCH_CHECK(scale.size(0) == zero_point.size(0),
```
nit: after the `dim() == 1` check, this can be `scale.numel()` == `zero_point.numel()`

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/fake_quantize_per_channel_affine.cpp:
```diff
@@ -0,0 +1,148 @@
+  TORCH_CHECK(scale.size(0) == self.size(axis),
+  "dimensions of scale and zero-point are not consistent with input tensor")
+
+
```
can do do clang-format?

dzhulgakov commented on aten/src/ATen/native/quantized/cuda/fake_quantize_per_channel_affine.cu:
```diff
@@ -0,0 +1,143 @@
+    auto Y_slice = Y.slice(axis,i,i+1);
+    float sc = scale[i].item().toFloat();
+    int64_t zp = zero_point[i].item().toLong();
+    fake_quantize_slice_cuda(Y_slice, X_slice, sc, zp, quant_min, quant_max);
```
you could declare this function with REGISTER_DISPATCH from DispatchStub.h and they you don't need to duplicate the top-level function at all. You'd just do REGISTER_CUDA_DISPATCH (and you'd move CPU implementation to per-architecture folder)

raghuramank100 replied:
Will mark as a TODO

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/fake_quantize_per_channel_affine.cpp:
```diff
@@ -0,0 +1,148 @@
+    float inv_scale = 1.0f / sc;
+    int64_t z_point = zero_point[i].item().toLong();
+    fake_quantize_slice(output_slice, input_slice,
+                           sc, z_point, quant_min, quant_max);
```
clang-format

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/fake_quantize_per_channel_affine.cpp:
```diff
@@ -0,0 +1,148 @@
+  TORCH_CHECK(dY.scalar_type() == ScalarType::Float);
+  TORCH_CHECK(X.scalar_type() == ScalarType::Float);
+
+  TORCH_CHECK(X.numel() == dY.numel(), "`X` and `dY` are not the same size");
```
this is checking number of elements, I think what you want is `X.sizes() == dY.sizes()`?

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/fake_quantize_per_channel_affine.cpp:
```diff
@@ -0,0 +1,148 @@
+      "`quant_min` should be less than or \
+        equal to `quant_max`.");
+
+  TORCH_CHECK(scale.size(0) == zero_point.size(0),
```
same comment as before

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/fake_quantize_per_tensor_affine.cpp:
```diff
@@ -2,7 +2,7 @@
 #include <ATen/native/TensorIterator.h>
 #include <ATen/native/cpu/Loops.h>
-
+#include "fake_quantize_core.h"
```
can you use full path here?

jerryzh168 commented on aten/src/ATen/native/quantized/cuda/fake_quantize_core.h:
```diff
@@ -0,0 +1,26 @@
+#pragma once
```
I think this file should be called `.cuh`

raghuramank100 replied:
No, it is not necessary. We should use cuh only when the file itself uses cuda primitives. In this case, the file does not do that .

jerryzh168 commented on aten/src/ATen/native/quantized/cuda/fake_quantize_per_channel_affine.cu:
```diff
@@ -0,0 +1,143 @@
+  TORCH_CHECK(dY.scalar_type() == ScalarType::Float);
+  TORCH_CHECK(X.scalar_type() == ScalarType::Float);
+
+  TORCH_CHECK(X.numel() == dY.numel(), "`X` and `dY` are not the same size");
```
same comment as before, please update accordinlgy

jerryzh168 commented on aten/src/ATen/native/quantized/cuda/fake_quantize_per_tensor_affine.cu:
```diff
@@ -2,28 +2,21 @@
 #include <ATen/NativeFunctions.h>
 #include <ATen/cuda/CUDAApplyUtils.cuh>
 #include <cmath>
+#include "fake_quantize_core.h"
```
same here, I think we should include full path

jerryzh168 commented on test/hypothesis_utils.py:
```diff
@@ -182,6 +182,38 @@ def tensor(draw, shapes=None, elements=None, qparams=None):
+        elements = st.floats(min_value, max_value, allow_infinity=False,
+                             allow_nan=False)
+    X = draw(stnp.arrays(dtype=np.float32, elements=elements, shape=_shape))
+    # Recompute the scale and zero_points according to the X statistics.
```
I don't really understand, why do we need to recalculate qparams?

raghuramank100 replied:
This is the same logic that is used for the per-tensor quantization, if we dont calculate qparams we end up with invalid choices of scale and zero-point that do not match the tensor. For example, you could end up with a tensor having a small dynamic range having a very large scale or even an invalid choice of zero-point.

jerryzh168 replied:
then we should change the tensor generation process to make sure zero_point is valid I think, generate tensors with some scale/zero_point and then recalculate the scale/zero_point again doesn't sound correct

raghuramank100 replied:
This is the same issue that @llyfacebook has been looking at. Let me address this in a seperate PR that fixes hypothesis_utils for both per-tensor and per-channel. (@llyfacebook : Do you have something in the works for this?)

llyfacebook replied:
Currently, I think we still need to do recalculate. Otherwise, all the resulted matrix will be with same element given large scale.

jerryzh168 commented on test/test_fake_quant.py:
```diff
@@ -8,37 +8,70 @@
+
+# Helper function used to simulate per-channel fake-quant against any axis
+def _permute_to_axis_zero(X, axis):
+    x_dim = X.dim()
```
nit: this can be removed since x_dim is used only once

dzhulgakov commented on torch/quantization/fake_quantize.py:
```diff
@@ -76,11 +77,17 @@ def calculate_qparams(self):
-            self.observer(X)
-            scale, zero_point = self.calculate_qparams()
-            self.scale, self.zero_point = float(scale), int(zero_point)
+            with torch.no_grad():
```
I think detach is better because no_grad is not scriptable

jerryzh168 commented on test/test_fake_quant.py:
```diff
@@ -8,37 +8,70 @@
     res[mask] = dY[mask]
     return res
 
+def to_tensor(self, X, device):
```
looks like this function is not used right now, but we should use it. can you use this to refactor places like L82?

jerryzh168 commented on test/test_fake_quant.py:
```diff
@@ -142,5 +175,119 @@ def test_fq_serializable(self):
         self.assertEqual(loaded_fq_module.calculate_qparams(), fq_module.calculate_qparams())
 
 
+class TestFakeQuantizePerChannel(TestCase):
```
Is it possible to merge this with TestFakeQuantizePerTensor rather than copy paste the code?

raghuramank100 replied:
The problem is with the hypothesis decorator outside the function. In one case it is a regular tensor and in another a per-channel tensor, not sure how to merge this.

jerryzh168 replied:
I think we can use draw to achieve this: https://hypothesis.readthedocs.io/en/latest/data.html#drawing-interactively-in-tests, but we can do this in a separate PR.

dzhulgakov commented on torch/quantization/observer.py:
```diff
@@ -115,7 +115,11 @@ def _calculate_per_channel_qparams(self, min_vals, max_vals):
+        # Needed to ensure that fake quantization module and operator level numerics match
+        # for test:
+        # caffe2/test:fake_quant - test_fq_module (test_fake_quant.TestFakeQuantizePerChannel)
+        scales = scales.to(torch.float)
```
you already initialize `scales` with torch.ones() above which is float. Why is it needed?

also remove the reference to the test - it's not helpful

dzhulgakov reviewed (approved):
Accepting to unblock, but there are comments inline to fix

raghuramank10000 committed:
Update on "Per channel fake quant"


Per-channel fake quant cpu and cuda operators,
per-channel support in fake quant module,
tests for per-channel fake-quant and serializability of fake quant modules


Differential Revision: [D17439406](https://our.internmc.facebook.com/intern/diff/D17439406/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Per channel fake quant"


Per-channel fake quant cpu and cuda operators,
per-channel support in fake quant module,
tests for per-channel fake-quant and serializability of fake quant modules


Differential Revision: [D17439406](https://our.internmc.facebook.com/intern/diff/D17439406/)

[ghstack-poisoned]

jerryzh168 reviewed (approved):
approve to unblock

raghuramank10000 committed:
Update on "Per channel fake quant"


Per-channel fake quant cpu and cuda operators,
per-channel support in fake quant module,
tests for per-channel fake-quant and serializability of fake quant modules


Differential Revision: [D17439406](https://our.internmc.facebook.com/intern/diff/D17439406/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Per channel fake quant"


Per-channel fake quant cpu and cuda operators,
per-channel support in fake quant module,
tests for per-channel fake-quant and serializability of fake quant modules


Differential Revision: [D17439406](https://our.internmc.facebook.com/intern/diff/D17439406/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Per channel fake quant"


Per-channel fake quant cpu and cuda operators,
per-channel support in fake quant module,
tests for per-channel fake-quant and serializability of fake quant modules


Differential Revision: [D17439406](https://our.internmc.facebook.com/intern/diff/D17439406/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Per channel fake quant"


Per-channel fake quant cpu and cuda operators,
per-channel support in fake quant module,
tests for per-channel fake-quant and serializability of fake quant modules


Differential Revision: [D17439406](https://our.internmc.facebook.com/intern/diff/D17439406/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Per channel fake quant"


Per-channel fake quant cpu and cuda operators,
per-channel support in fake quant module,
tests for per-channel fake-quant and serializability of fake quant modules


Differential Revision: [D17439406](https://our.internmc.facebook.com/intern/diff/D17439406/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Per channel fake quant"


Per-channel fake quant cpu and cuda operators,
per-channel support in fake quant module,
tests for per-channel fake-quant and serializability of fake quant modules


Differential Revision: [D17439406](https://our.internmc.facebook.com/intern/diff/D17439406/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Per channel fake quant"


Per-channel fake quant cpu and cuda operators,
per-channel support in fake quant module,
tests for per-channel fake-quant and serializability of fake quant modules


Differential Revision: [D17439406](https://our.internmc.facebook.com/intern/diff/D17439406/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Per channel fake quant"


Per-channel fake quant cpu and cuda operators,
per-channel support in fake quant module,
tests for per-channel fake-quant and serializability of fake quant modules


Differential Revision: [D17439406](https://our.internmc.facebook.com/intern/diff/D17439406/)

[ghstack-poisoned]
